---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 03-nanopore-basecalling.md in _episodes_rmd/
title: "Nanopore - Basecalling"
teaching: 10
exercises: 10
questions:
- "How to write a lesson using RMarkdown?"
objectives:
- "Explain how to use RMarkdown with the new lesson template."
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
source: Rmd
---



> possible example dataset of E coli available from https://zenodo.org/record/1257429



## FAST5 / HDF5 data

 - Each pore produces a HUGE amount of data - very roughly, 1Gbp of sequence data requires 1GB of storage (e.g., as gzipped fastq), but to generate 1Gbp of sequence requires 10GB of electrical trace data, so potentially 500GB of files for a 72 hour MinION run.
 - The electrical trace data is saved as ".fast5", which utises the HDF5 file format:
 
"Hierarchical Data Format (HDF) is a set of file formats (HDF4, HDF5) designed to store and organize large amounts of data. Originally developed at the National Center for Supercomputing Applications, it is supported by The HDF Group, a non-profit corporation whose mission is to ensure continued development of HDF5 technologies and the continued accessibility of data stored in HDF."

### https://www.neonscience.org/resources/learning-hub/tutorials/about-hdf5



## Nanopore workflow



<img src="../fig/01-nanopore-workflow.png" alt="plot of chunk unnamed-chunk-2" height="180" style="display: block; margin: auto;" />

 - ONT provides software (MinKNOW) for operating the MinION, and for generating the sequence data (e.g., the `guppy` basecaller).
 - Once the raw FAST5 data have been converted to basecalls, we can use more familiar tools for quality assessment and analysis (e.g., FastQC).

[https://nanoporetech.com/nanopore-sequencing-data-analysis](https://nanoporetech.com/nanopore-sequencing-data-analysis)



## Basecalling: `guppy`

 - `guppy` is a neural network based basecaller.
    - analyses the electrical trace data and predicts base 
    - it is GPU-aware, and can basecall in real time 
    - can also call base modifications (e.g., 5mC, 6mA)
    - high accuracy mode (slower) and super-high accuracy mode (even slower) can improve basecalls post-sequencing
 - The software is provided by ONT, and the source code is available via a developer agreement.
 - MANY other machine learning basecallers have been proposed.
 - Output is the standard "FASTQ" format for sequence data.


~~~
need the commands for running guppy
~~~
{: .bash}

<BR><BR><BR><BR><BR><BR>
# Recap: sequence data



## Assessing sequence quality: phred scores



<img src="../fig/01-phred.png" alt="plot of chunk unnamed-chunk-3" height="450" style="display: block; margin: auto;" />

### Ewing B, Green P. (1998): Base-calling of automated sequencer traces using phred. II. Error probabilities. Genome Res. 8(3):186-194.



## Assessing seqeunce quality: phred scores




<img src="../fig/01-phred2.png" alt="plot of chunk unnamed-chunk-4" height="300" style="display: block; margin: auto;" />

Can use ASCII to represent quality scores by adding 33 to the phred score and converting to ASCII.
 - Quality score of 38 becomes 38+33=71: “G” 

[http://en.wikipedia.org/wiki/Phred_quality_score](http://en.wikipedia.org/wiki/Phred_quality_score)


 


<img src="../fig/01-asciitable.png" alt="plot of chunk unnamed-chunk-5" height="600" style="display: block; margin: auto;" />




<img src="../fig/01-phredtrace.png" alt="plot of chunk unnamed-chunk-6" height="550" style="display: block; margin: auto;" />

[http://en.wikipedia.org/wiki/Phred_quality_score](http://en.wikipedia.org/wiki/Phred_quality_score)



## Sequence data format

 - The FASTQ format allows the storage of both sequence and quality information for each read.
 - This is a compact text-based format that has become the de facto standard for storing data from next generation sequencing experiments.



## Fastq format 





<img src="../fig/01-fastq1.png" alt="plot of chunk unnamed-chunk-7" height="420" style="display: block; margin: auto;" />

### http://en.wikipedia.org/wiki/FASTQ_format

*** =right
 - Line 1: '@' character followed by sequence identifier and optional description.
 - Line 2: base calls.
 - Line 3: '+' character, ptionally followed by the same sequence identifier (and description) again.
 - Line 4: quality scores



## Fastq format



<img src="../fig/01-fastq2.png" alt="plot of chunk unnamed-chunk-8" height="475" style="display: block; margin: auto;" />

### http://en.wikipedia.org/wiki/FASTQ_format



## Quality Assessment

 - The information in the FASTQ file can be used to assess the read quality.
 - A number of tools are available on the unix/linux command line, and/or within R for manipulating FASTQ data.



## Sequence alignment

 - If a “reference” genome exists for the organism you are sequencing, reads can be “aligned” to the reference.
 - This involves finding the place in the reference genome that each read matches to.
 - Due to high sequence similarity within members of the same species, most reads should map to the reference.



## Tools for generating alignments

 - There are MANY software packages available for aligning data from next generation sequencing experiments.
 - The two "original" short read aligners were:
    - BWA: http://bio-bwa.sourceforge.net
    - Bowtie: http://bowtie-bio.sourceforge.net
 - More recently MANY more "aligners" have been developed
 - For long-read data, ONT provides the Minimap2 aligner (but there are also a number of other options).



~~~
minimap2
~~~
{: .bash}



## Existing tutorials 

https://labs.epi2me.io/nbindex/

- https://labs.epi2me.io/notebooks/Introduction_to_Fast5_files.html
- https://labs.epi2me.io/notebooks/Basic_QC_Tutorial.html

https://denbi-nanopore-training-course.readthedocs.io/en/latest/data.html
