---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 04-nanopore-qc.md in _episodes_rmd/
title: "Nanopore: quality assessment"
teaching: 10
exercises: 10
questions:
- "How can I perform quality assessment of ONT data?"
objectives:
- "Assess the quality of reads."
- "Generate summary metrics."
keypoints:
- "FASTQC and NanoPore provide pre-formatted reports on run characteristics and read quality."
- "Bioawk can be used to generate summary metrics and ask specific questions about the read data."
source: Rmd
---



## Software for quality assessment

In this part of the module, we'll look at software applications that can be used to assess the quality of data from ONT sequencing runs. We'll be focussing on two pieces of software: FASTQC and NanoPlot.

- FASTQC (you've already seen this): generic tool for assessing the quality of next generation sequencing data.
- NanoPlot: quality assessment software specifically built for ONT data.

### NanoPlot

Let's use the NanoPlot software to assess the quality of our `.fastq` data.  We can run NanoPlot interactively (i.e., we don't need to submit it as a slurm job), and it wil produce an HTML-based report that can be viewed in a browser.


```
# Load NanoPlot module
module load NanoPlot
```
 
```
# Run NanoPlot on passed fastq data
NanoPlot -o nanoplot_fastmodel_subset/ \
         -p ecoli_fastmodel_subset_ \
         --fastq fastq_fastmodel_subset/pass/fastq_runid_*.gz
```

```
# create single .fastq.gz file containing all reads 
# NB: not actually needed for bioawk  - can just use: 
#     fastq_fastmodel_subset/pass/*gz
#     Needed for FastQC though

cat fastq_fastmodel_subset/pass/*.gz > ecoli-pass.fastq.gz
zcat ecoli-pass.fastq.gz | wc -l 
echo 48924/4 | bc
```

```
module load FastQC

mkdir fastqc_fastmodel_subset
# NB: I get a java memory error if I don't use two threads
#     FastQC allocates 250Mb per thread. Can't remember how 
#     to specify more memory for java...
fastqc -t 2 -o fastqc_fastmodel_subset ecoli-pass.fastq.gz
```

```
module load bioawk

# Extract read name and length
bioawk -c fastx '{print $name,length($seq)}' ecoli-pass.fastq.gz | head

# Number of reads
bioawk -c fastx '{print $name,length($seq)}' ecoli-pass.fastq.gz | wc -l

# Total number of bases in reads (subset to demonstrate)
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | head 
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | head | paste -sd+ 
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | head | paste -sd+ | bc

# Total number of bases in reads 
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | \
  paste -sd+ | bc
  
# Number of reads longer than 10000 bases
bioawk -c fastx '{print (length($seq) > 10000)}' ecoli-pass.fastq.gz | head
bioawk -c fastx '{print (length($seq) > 5000)}' ecoli-pass.fastq.gz | \
  paste -sd+ | bc

# NB: mean quality scores aren't correct
# Maybe skip this stuff...
bioawk -c fastx '{print $name,length($seq),meanqual($seq)}' ecoli-pass.fastq.gz | head
```

