---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 04-nanopore-qc.md in _episodes_rmd/
title: "Nanopore - QC"
teaching: 10
exercises: 10
questions:
- "How can I perform QC on nanopore data?"
objectives:
- "Assess the quality of reads."
- "Generate summary metrics"
keypoints:
- "Bioawk can be used to generate summary metrics"
source: Rmd
---




## Assessing sequence quality: phred scores



<img src="../fig/01-phred.png" alt="plot of chunk unnamed-chunk-2" height="450" style="display: block; margin: auto;" />

### Ewing B, Green P. (1998): Base-calling of automated sequencer traces using phred. II. Error probabilities. Genome Res. 8(3):186-194.



## Assessing seqeunce quality: phred scores




<img src="../fig/01-phred2.png" alt="plot of chunk unnamed-chunk-3" height="300" style="display: block; margin: auto;" />

Can use ASCII to represent quality scores by adding 33 to the phred score and converting to ASCII.
 - Quality score of 38 becomes 38+33=71: “G” 

[http://en.wikipedia.org/wiki/Phred_quality_score](http://en.wikipedia.org/wiki/Phred_quality_score)


 


<img src="../fig/01-asciitable.png" alt="plot of chunk unnamed-chunk-4" height="600" style="display: block; margin: auto;" />




<img src="../fig/01-phredtrace.png" alt="plot of chunk unnamed-chunk-5" height="550" style="display: block; margin: auto;" />

[http://en.wikipedia.org/wiki/Phred_quality_score](http://en.wikipedia.org/wiki/Phred_quality_score)



## Sequence data format

 - The FASTQ format allows the storage of both sequence and quality information for each read.
 - This is a compact text-based format that has become the de facto standard for storing data from next generation sequencing experiments.



## Fastq format 





<img src="../fig/01-fastq1.png" alt="plot of chunk unnamed-chunk-6" height="420" style="display: block; margin: auto;" />

### http://en.wikipedia.org/wiki/FASTQ_format

*** =right
 - Line 1: '@' character followed by sequence identifier and optional description.
 - Line 2: base calls.
 - Line 3: '+' character, ptionally followed by the same sequence identifier (and description) again.
 - Line 4: quality scores



## Fastq format



<img src="../fig/01-fastq2.png" alt="plot of chunk unnamed-chunk-7" height="475" style="display: block; margin: auto;" />

### http://en.wikipedia.org/wiki/FASTQ_format



## Quality Assessment

 - The information in the FASTQ file can be used to assess the read quality.
 - A number of tools are available on the unix/linux command line, and/or within R for manipulating FASTQ data.



## NanoPlot / FastQC / bioawk

Code for e. coli data set on NeSI:

```
# Load NanoPlot module
module load NanoPlot
```
 
```
# Run NanoPlot on passed fastq data
NanoPlot -o nanoplot_fastmodel_subset/ \
         -p ecoli_fastmodel_subset_ \
         --fastq fastq_fastmodel_subset/pass/fastq_runid_*.gz
```

```
# create single .fastq.gz file containing all reads 
# NB: not actually needed for bioawk  - can just use: 
#     fastq_fastmodel_subset/pass/*gz
#     Needed for FastQC though

cat fastq_fastmodel_subset/pass/*.gz > ecoli-pass.fastq.gz
zcat ecoli-pass.fastq.gz | wc -l 
echo 48924/4 | bc
```

```
module load FastQC

mkdir fastqc_fastmodel_subset
# NB: I get a java memory error if I don't use two threads
#     FastQC allocates 250Mb per thread. Can't remember how 
#     to specify more memory for java...
fastqc -t 2 -o fastqc_fastmodel_subset ecoli-pass.fastq.gz
```

```
module load bioawk

# Extract read name and length
bioawk -c fastx '{print $name,length($seq)}' ecoli-pass.fastq.gz | head

# Number of reads
bioawk -c fastx '{print $name,length($seq)}' ecoli-pass.fastq.gz | wc -l

# Total number of bases in reads (subset to demonstrate)
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | head 
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | head | paste -sd+ 
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | head | paste -sd+ | bc

# Total number of bases in reads 
bioawk -c fastx '{print length($seq)}' ecoli-pass.fastq.gz | \
  paste -sd+ | bc
  
# Number of reads longer than 10000 bases
bioawk -c fastx '{print (length($seq) > 10000)}' ecoli-pass.fastq.gz | head
bioawk -c fastx '{print (length($seq) > 5000)}' ecoli-pass.fastq.gz | \
  paste -sd+ | bc

# NB: mean quality scores aren't correct
# Maybe skip this stuff...
bioawk -c fastx '{print $name,length($seq),meanqual($seq)}' ecoli-pass.fastq.gz | head
```

