---
title: "Nanopore - Variant Calling"
teaching: 10
exercises: 10
questions:
- "How to write a lesson using RMarkdown?"
objectives:
- "Explain how to use RMarkdown with the new lesson template."
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("05-")
```

> possible example dataset of E coli available from https://zenodo.org/record/1257429






## Sequence alignment

 - If a “reference” genome exists for the organism you are sequencing, reads can be “aligned” to the reference.
 - This involves finding the place in the reference genome that each read matches to.
 - Due to high sequence similarity within members of the same species, most reads should map to the reference.



## Tools for generating alignments

 - There are MANY software packages available for aligning data from next generation sequencing experiments.
 - The two "original" short read aligners were:
    - BWA: http://bio-bwa.sourceforge.net
    - Bowtie: http://bowtie-bio.sourceforge.net
 - More recently MANY more "aligners" have been developed
 - For long-read data, ONT provides the Minimap2 aligner (but there are also a number of other options).



~~~
minimap2
~~~
{: .bash}



## Existing tutorials 

https://labs.epi2me.io/nbindex/

- https://labs.epi2me.io/notebooks/Introduction_to_Fast5_files.html
- https://labs.epi2me.io/notebooks/Basic_QC_Tutorial.html

https://denbi-nanopore-training-course.readthedocs.io/en/latest/data.html